### PR TITLE
Simplify local builds with Docker Compose, Also updated README

### PR DESCRIPTION
**Pull request by jarrodu**
Feel free to look around if you like. I will get back to this tonight or tomorrow. :-)

The main things I still need to do is to provide a reason for the change and update the readme with how to use it.

### CODE REVIEW

**Comment by jarrodwb** [comment]
I still need to update the readme but could some of you check this out?

I am using docker and docker-compose to make it easier for current developers to stay in sync and also make it easier for new contributions. Scala developers probably don't want to deal with the specifics of ruby, rbenv, rvm, bundle, jekyll, etc.

They are probably familiar with html, markdown, docker, and docker-compose.

The commands to run are in the bin directory but should all be run from the root directory. ie. `bin/serve` or `bin/bundle exec ...`

The first time  you run `bin/serve` a docker image will be downloaded and all the gems will also be downloaded. Then jekyll will build the site and start the server. If  you run it again you will already have the image and the gems. 

Also checkout:
`cat bin/*`

If this PR gets any traction or feedback I will polish it up and start working on the readme. :-)

@heathermiller @SethTisue @ashawley @OlivierBlanvillain @Philippus

**Comment by jarrodu** [comment]
Is anyone interested in trying this out? If not than I can close the PR.

I am hoping this will help get more contributers into the project.

**Comment by pfgray** [comment]
@jarrodu , FWIW I had not known of this project until 30 minutes ago, and got this up and running, and made a change in under 10 mins:

screen shot 2018-02-20 at 2 05 06 pm

**Comment by jarrodu** [comment]
That is so cool! Thanks so much for sharing.

**Comment by propensive** [comment]
I tried it, and it works! I had to install `docker-compose` (a quick check to see whether it exists in the `bin/serve` script would be useful), but the process of testing the website was much easier that my previous experience, so I would like to see this merged.

If you would like to finish polishing up the PR, I'll be happy to merge it unless anyone else objects? I trust that everything you've included here is complementary to the existing process for updating the website. In the unlikely event that it does, make sure it's documented as anyone with an existing process which works for them would certainly need to know! :)

**Comment by jarrodu** [comment]
Thanks for the positive encouragement. I will go through everything in detail tomorrow. Today I am getting a cat! I will also review and update the README.

As far as I can tell this should not change the build process for people that are already comfortable with their setups. It will also not change the Travis build. We will see how well people like this and maybe in the future I can expand the scope a bit.

**Comment by SethTisue** [comment]
I haven't had the chance to try this yet, but it certainly looks like an appealing option to offer.

File: README.md [hunk 1]
```diff
@@ -103,31 +53,29 @@ You can use these fields in the YAML front matter later in your document. For ex
...
 ### Linking to internal pages
 
-The least error-prone way to link between documents, to link to local images, or anything else: `[link text]({{ site.baseurl }}/path/to/page/page.html)`
+The least error-prone way to make links is to use this format: `[link text]({{ site.baseurl }}/path/to/page/page.html)`
 
-Here, `{{ site.baseurl }}` is a site-wide variable that represents the root directory of the static site. So, to display the Scala logo image, located in `img/scala-logo.png`, one must simply write: `![Img alt text]({{ site.baseurl }}/resources/img/scala-logo.png)`
+`{{ site.baseurl }}` is a site-wide variable that represents the root directory of the static site. So, to display the Scala logo image you can simply write: `![Img alt text]({{ site.baseurl }}/resources/img/scala-logo.png)`
 
-## Permalinks
+### Permalinks
```
Comment (jarrodu) [inline README.md]: Does anyone know how to fix this rendering issue? It is also present on the master branch.

Comment (Philippus) [reply README.md]: Looks like the character between '##' and 'Permalinks' is not a space.

**Comment by ashawley** [comment]
I'm sure they are interdependent, but it seems like there are at least 3 changes under consideration here:

1. Upgrade the gems
2. Add docker
3. Improve the README

In the interest of getting this important work integrated rather than delayed, would it be possible to make the changes separately?

I think I have an old version of docker-compose, so it didn't work for me.  I'll need to look in to upgrading.

```
$ ./bin/serve
WARNING: Please be aware that .yml is the expected extension in most cases, and using .yaml can cause compatibility issues in future.

ERROR: In file './docker-compose.yaml' service 'version' doesn't have any configuration options. All top level keys in your docker-compose.yml must map to a dictionary of configuration options.
```

File: README.md (hunk 1 above)
Comment (jarrodu) [reply README.md]: Yup. That fixed it. :-)

**Comment by jarrodu** [comment]
I minimized the differences between master's and docker-compose's version of Gemfile.lock. This should make reviewing easier. Since my goal here is to make it easier and more inviting for new contributors I prefer to keep the rest of the changes in this PR. The existing contributors can continue using the workflows they are used to.

I am running Compose 1.8 which is the default if you are running Ubuntu 17.10. Version 1.19 installed using pip also works.

**Comment by ashawley** [comment]
> I minimized the differences between master's and docker-compose's version of Gemfile.lock.

Thanks, @jarrodu. That will make it easier to review.

> my goal here is to make it easier and more inviting for new contributors

I'm all for this.  I tried to do a small piece of that in #693, which I suspect is one of the technical hurdles that motivates using docker actually.

File: README.md [hunk 2]
```diff
@@ -103,31 +53,29 @@ You can use these fields in the YAML front matter later in your document. For ex
...
+`{{ site.baseurl }}` is a site-wide variable that represents the root directory of the static site. So, to display the Scala logo image you can simply write: `![Img alt text]({{ site.baseurl }}/resources/img/scala-logo.png)`
 
-## Permalinks
+### Permalinks
 
-In this new version of the scala-lang site we've tried to follow a `pretty permalink` style, so that any generated page will have an permalink finishing in a slash character (`/`). This will tell Jekyll to build that particular page as an `index.html` inside a folder with a name as specified in the provided permalink. i.e.: if a page has a permalink as follows:
+We trie to follow a [pretty permalink](https://jekyllrb.com/docs/permalinks/) style, so that any generated page will have a link finishing in a slash character (`/`). This will tell Jekyll to build that particular page as an `index.html` inside a folder with a name as specified in the provided permalink. i.e.: if a page has a permalink as follows:
```
Comment (propensive) [inline README.md]: We *try*

**Comment by propensive** [comment]
Thank you again, @jarrodu! Happy to have this contribution. :)

**Pull request merged**